app/src/org/commcare/tasks/DataPullTask.java

```diff
@@ -102,258 +101,296 @@ public DataPullTask(String username, String password,
         if (!CommCareSessionService.sessionAliveLock.tryLock()) {
-            // TODO PLM: once this task is refactored into manageable
-            // components, it should use the ManagedAsyncTask pattern of
-            // checking for isCancelled() and aborting at safe places.
-            return new ResultAndError<>(PullTaskResult.UNKNOWN_FAILURE, "Cannot sync while a logout is in process");
+            // Don't try to sync if logging out is occurring
+            return new ResultAndError<>(PullTaskResult.UNKNOWN_FAILURE,
+                    "Cannot sync while a logout is in process");
+        }
+        try {
+            return doTaskBackgroundHelper();
+        } finally {
+            CommCareSessionService.sessionAliveLock.unlock();
+        }
+    }
+
+    private ResultAndError<PullTaskResult> doTaskBackgroundHelper() {
+        publishProgress(PROGRESS_STARTED);
+        recordSyncAttemptTime();
+        Logger.log(AndroidLogger.TYPE_USER, "Starting Sync");
+        determineIfLoginNeeded();
+
+        AndroidTransactionParserFactory factory = getTransactionParserFactory();
+        byte[] wrappedEncryptionKey = getEncryptionKey();
+        if (wrappedEncryptionKey == null) {
+            this.publishProgress(PROGRESS_DONE);
+            return new ResultAndError<>(PullTaskResult.UNKNOWN_FAILURE,
+                    "Unable to get or generate encryption key");
+        } else {
+            factory.initUserParser(wrappedEncryptionKey);
         }
 
-        // Wrap in a 'try' to enable a 'finally' close that releases the
-        // sessionAliveLock.
+        if (isCancelled()) {
+            // Avoid making http request if user cancelled the task (NOTE: In this case, the result
+            // returned is never processed since cancelled task results are sent to onCancelled)
+            return new ResultAndError<>(PullTaskResult.UNKNOWN_FAILURE, "");
+        }
+
+        PullTaskResult responseError = PullTaskResult.UNKNOWN_FAILURE;
         try {
-            publishProgress(PROGRESS_STARTED);
-            CommCareApp app = CommCareApplication._().getCurrentApp();
-            SharedPreferences prefs = app.getAppPreferences();
-
-            mTotalItems = -1;
-            mCurrentProgress = -1;
-
-            //Whether or not we should be generating the first key
-            String keyServer = CommCarePreferences.getKeyServer();
-            boolean useExternalKeys = !(keyServer == null || keyServer.equals(""));
-
-            boolean loginNeeded = true;
-            boolean useRequestFlags = false;
-            try {
-                loginNeeded = !CommCareApplication._().getSession().isActive();
-            } catch (SessionUnavailableException sue) {
-                //expected if we aren't initialized.
+            ResultAndError<PullTaskResult> resultFromRequest = makeRequestAndHandleResponse(factory);
+            if (resultFromRequest != null) {
+                return resultFromRequest;

```

Comment:
This is a kind of odd pattern that I introduced in order to preserve the precise behavior in the original code, in terms of when an error result was returned. In the old code, some errors would return a failure result immediately, while others would be allowed to fall through to the code on lines 174-176 below. Since I couldn't tell if that original design was intentional or incidental, I didn't want to change it. So any errors that were falling through to those lines in the old code now return null in their helper methods so that it will fall through here.